cmd/cli/kubectl-kyverno/commands/apply/print.go

```diff
@@ -0,0 +1,152 @@
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+	"time"
+
+	kyvernov1 "github.com/kyverno/kyverno/api/kyverno/v1"
+	kyvernov2 "github.com/kyverno/kyverno/api/kyverno/v2"
+	kyvernov2beta1 "github.com/kyverno/kyverno/api/kyverno/v2beta1"
+	"github.com/kyverno/kyverno/api/policyreport/v1alpha2"
+	"github.com/kyverno/kyverno/cmd/cli/kubectl-kyverno/processor"
+	"github.com/kyverno/kyverno/cmd/cli/kubectl-kyverno/report"
+	engineapi "github.com/kyverno/kyverno/pkg/engine/api"
+	kyvernoreports "github.com/kyverno/kyverno/pkg/utils/report"
+	"github.com/opentracing/opentracing-go/log"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/yaml"
+)
+
+func printSkippedAndInvalidPolicies(out io.Writer, skipInvalidPolicies SkippedInvalidPolicies) {
+	if len(skipInvalidPolicies.skipped) > 0 {
+		fmt.Fprintln(out, divider)
+		fmt.Fprintln(out, "Policies Skipped (as required variables are not provided by the user):")
+		for i, policyName := range skipInvalidPolicies.skipped {
+			fmt.Fprintf(out, "%d. %s\n", i+1, policyName)
+		}
+		fmt.Fprintln(out, divider)
+	}
+	if len(skipInvalidPolicies.invalid) > 0 {
+		fmt.Fprintln(out, divider)
+		fmt.Fprintln(out, "Invalid Policies:")
+		for i, policyName := range skipInvalidPolicies.invalid {
+			fmt.Fprintf(out, "%d. %s\n", i+1, policyName)
+		}
+		fmt.Fprintln(out, divider)
+	}
+}
+
+func printReports(out io.Writer, engineResponses []engineapi.EngineResponse, auditWarn bool) {
+	clustered, namespaced := report.ComputePolicyReports(auditWarn, engineResponses...)
+	if len(clustered) > 0 {
+		report := report.MergeClusterReports(clustered)
+		yamlReport, _ := yaml.Marshal(report)
+		fmt.Fprintln(out, string(yamlReport))
+	}
+	for _, r := range namespaced {
+		fmt.Fprintln(out, string("---"))
+		yamlReport, _ := yaml.Marshal(r)
+		fmt.Fprintln(out, string(yamlReport))
+	}
+}
+
+func printExceptions(out io.Writer, engineResponses []engineapi.EngineResponse, auditWarn bool, ttl time.Duration) {
+	clustered, _ := report.ComputePolicyReports(auditWarn, engineResponses...)
+	for _, report := range clustered {
+		for _, result := range report.Results {
+			if result.Result == "fail" {
+				if err := printException(out, result, ttl); err != nil {
+					log.Error(err)
+				}
+			}
+		}
+	}
+}
+
+func printException(out io.Writer, result v1alpha2.PolicyReportResult, ttl time.Duration) error {
+	for _, r := range result.Resources {
+		name := strings.Join([]string{result.Policy, result.Rule, r.Namespace, r.Name}, "-")
+		exception := kyvernov2.PolicyException{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       "PolicyException",
+				APIVersion: kyvernov2.SchemeGroupVersion.String(),
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: name,
+			},
+			Spec: kyvernov2.PolicyExceptionSpec{
+				Match: kyvernov2beta1.MatchResources{
+					All: kyvernov1.ResourceFilters{
+						kyvernov1.ResourceFilter{
+							ResourceDescription: kyvernov1.ResourceDescription{
+								Kinds:      []string{r.Kind},
```

Comment:
Does it matter in the CLI that the exception is only against a single kind and name?

If it generates an exception for a deployment the related ReplicaSet and Pod are still failing because the PolicyException has no effect.

Reply:
@fjogeleit - good point. Yes, it seems like we should generate exceptions for the replicaset and pods as well.

Reply:
Users need to create exceptions and match pod controllers and also include the autogenerated rule name.
```
exceptions:
  - policyName: disallow-host-namespaces
    ruleNames:
    - host-namespaces
    - autogen-host-namespaces
  match:
    any:
    - resources:
        kinds:
        - Pod
        - Deployment
```

Reply:
Thanks @MariamFahmy98! Ok, I tested with an exception similar to that and it works:

```yaml
---
apiVersion: kyverno.io/v2beta1
kind: PolicyException
metadata:
  labels:
  name: require-run-as-nonroot-autogen-run-as-non-root-default-nginx
spec:
  exceptions:
  - policyName: require-run-as-nonroot
    ruleNames:
    - run-as-non-root
    - autogen-run-as-non-root
  match:
    all:
    - resources:
        kinds:
        - Pod
        - Deployment
        namespaces:
        - default
        names:
        - "nginx"
        - "nginx-*"

```

It does not require `replicasets` in the exception. Do you know why?

Reply:
Did you also check for cronjob/job? Which might also be needed.

Reply:
add a check for cronjob/job